staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/controller.go

```diff
@@ -0,0 +1,200 @@
+		return
+	}
+
+	// only start one worker thread since its a slow moving API
+	go wait.Until(c.runWorker, time.Second, stopCh)
+
+	<-stopCh
+}
+
+func (c *Controller) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Controller) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	err := c.syncFn(key.(schema.GroupVersionKind))
+	if err == nil {
+		c.queue.Forget(key)
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("%v failed with: %v", key, err))
+	c.queue.AddRateLimited(key)
+	return true
+}
+
+func (c *Controller) sync(gvk schema.GroupVersionKind) error {
+	if c.openAPIAggregationManager == nil || !utilfeature.DefaultFeatureGate.Enabled(apiextensionsfeatures.CustomResourceValidation) {
+		return nil
+	}
+
+	crds, err := c.crdLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+	specName := gvk.String()
+	found := false
+
+	for _, crd := range crds {
+		if !apiextensions.IsCRDConditionTrue(crd, apiextensions.Established) {
+			continue
+		}
+		if crd.Spec.Group != gvk.Group || crd.Spec.Names.Kind != gvk.Kind {
+			continue
+		}
+		if !hasVersionServed(crd, gvk.Version) {
+			continue
+		}
+
+		found = true
+		spec, etag, err := BuildSwagger(crd, gvk.Version)
+		if err != nil {
+			return err
+		}
+		if err := c.openAPIAggregationManager.AddUpdateLocalAPIServiceSpec(specName, spec, etag); err != nil {
+			return err
+		}
+	}
+
+	if !found {
+		return c.openAPIAggregationManager.RemoveAPIServiceSpec(specName)
+	}
+	return nil
+}
+
+func (c *Controller) addCustomResourceDefinition(obj interface{}) {
+	castObj := obj.(*apiextensions.CustomResourceDefinition)
+	klog.V(4).Infof("Adding customresourcedefinition %s", castObj.Name)
+	c.enqueue(castObj)
+}
+
+func (c *Controller) updateCustomResourceDefinition(oldObj, newObj interface{}) {
+	castNewObj := newObj.(*apiextensions.CustomResourceDefinition)
+	castOldObj := oldObj.(*apiextensions.CustomResourceDefinition)
+	klog.V(4).Infof("Updating customresourcedefinition %s", castOldObj.Name)
+	c.enqueue(castNewObj)
+	c.enqueue(castOldObj)
+}
+
+func (c *Controller) deleteCustomResourceDefinition(obj interface{}) {
+	castObj, ok := obj.(*apiextensions.CustomResourceDefinition)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.Errorf("Couldn't get object from tombstone %#v", obj)
+			return
+		}
+		castObj, ok = tombstone.Obj.(*apiextensions.CustomResourceDefinition)
+		if !ok {
+			klog.Errorf("Tombstone contained object that is not expected %#v", obj)
+			return
+		}
+	}
+	klog.V(4).Infof("Deleting customresourcedefinition %q", castObj.Name)
+	c.enqueue(castObj)
+}
+
+func (c *Controller) enqueue(obj *apiextensions.CustomResourceDefinition) {
+	for _, v := range obj.Spec.Versions {
+		c.queue.Add(schema.GroupVersionKind{Group: obj.Spec.Group, Version: v.Name, Kind: obj.Spec.Names.Kind})
```

Comment:
can't we enqueue the GVR instead? It maps 1:1 to the CRD name and makes quite some code here simpler.

Reply:
I think it's the same question for enqueueing old object. We need to know what versions were there in the old CRD

Reply:
If we sync GroupResources and also register GroupResource in the aggregation manager, the whole group resource with all versions is replaced in the OpenAPI spec. Then we don't need this loop.